app/controllers/CampaignApi.scala

```diff
@@ -2,93 +2,90 @@ package controllers
 
 import com.gu.googleauth.AuthAction
 import model._
-import model.command._
-import model.reports._
 import play.api.Logger
-import play.api.libs.json.Json._
-import play.api.libs.json._
 import play.api.mvc._
 import repositories._
 import services.CampaignService
+import cats.syntax.either._
+import model.command._
+import play.api.libs.json.Json
 
 class CampaignApi(components: ControllerComponents, authAction: AuthAction[AnyContent])
   extends AbstractController(components) {
 
   def getCampaign(id: String) = authAction {
-    CampaignRepository.getCampaign(id) map { c =>
-      Ok(Json.toJson(c))
+    CampaignRepository.getCampaign(id) map { campaign =>
+      Ok(Json.toJson(campaign))
     } getOrElse NotFound
   }
 
   def getAllCampaigns() = authAction {
-    Ok(Json.toJson(CampaignRepository.getAllCampaigns()))
+    CampaignRepository.getAllCampaigns() match {
+      case Left(_)          => InternalServerError
+      case Right(campaigns) => Ok(Json.toJson(campaigns))
+    }
   }
 
   def getLatestCampaignAnalytics() = authAction {
-    Ok(Json.toJson(CampaignService.getLatestCampaignAnalytics()))
+    CampaignService.getLatestCampaignAnalytics() match {
+      case Left(JsonParsingError(error)) => InternalServerError(error)
+      case Left(_)                       => InternalServerError
+      case Right(analytics)              => Ok(Json.toJson(analytics))
+    }
   }
 
   def getLatestAnalyticsForCampaign(campaignId: String) = authAction {
-    Ok(Json.toJson(CampaignService.getLatestAnalyticsForCampaign(campaignId)))
+    CampaignService.getLatestAnalyticsForCampaign(campaignId) match {
+      case Left(LatestCampaignAnalyticsItemNotFound(error)) => NotFound(error)
+      case Left(CampaignNotFound(error))                    => NotFound(error)
+      case Left(JsonParsingError(error))                    => InternalServerError(error)
+      case Left(_)                                          => InternalServerError
+      case Right(analytics)                                 => Ok(Json.toJson(analytics))
+    }
   }
 
   def updateCampaign(id: String) = authAction { req =>
     req.body.asJson.flatMap(_.asOpt[Campaign]) match {
       case None => BadRequest("Could not convert json to campaign")
       case Some(campaign) =>
-        CampaignRepository.putCampaign(campaign)
-        Ok(Json.toJson(campaign))
+        CampaignRepository.putCampaign(campaign) match {
+          case Left(_)  => InternalServerError
+          case Right(_) => Ok(Json.toJson(campaign))
+        }
     }
   }
 
   def deleteCampaign(id: String) = authAction { _ =>
-    CampaignContentRepository.deleteContentForCampaign(id)
-    CampaignRepository.deleteCampaign(id)
-    NoContent
-  }
+    val result = for {
+      _ <- CampaignContentRepository.deleteContentForCampaign(id)
+      _ <- CampaignRepository.deleteCampaign(id)
+    } yield NoContent
```

Comment:
What happens in the case of `Left[CampaignCentralApiError]`? It isn't a valid return type for a controller, so now sure how this is compiling?

Reply:
Oops! Ok, I see the `getOrElse`.